File: backend/app/api/v1/team.py
"""Repair team endpoints.

A crew member sees only their own team's work. The team is taken from the
authenticated user's ``team_id``, never from a request parameter, so one crew
cannot read or act on another's jobs by guessing an id.
"""

from __future__ import annotations

import uuid
from datetime import UTC, datetime, timedelta

from fastapi import APIRouter, Depends, File, Form, Query, UploadFile
from sqlalchemy.orm import Session

from app.api.deps import get_session, require_team
from app.api.serializers import to_task
from app.core.config import settings
from app.core.enums import AssignmentStatus, PriorityLevel, UserRole
from app.core.errors import PermissionDeniedError, ValidationError
from app.core.rate_limit import write_rate_limit
from app.models.repair import RepairAssignment
from app.models.user import User
from app.repositories.repair import RepairRepository
from app.schemas.admin import (
    RepairTeamResponse,
    TaskResponse,
    TeamDashboardResponse,
)
from app.services.complaints import ComplaintService

router = APIRouter(prefix="/team", tags=["team"])


def _resolve_team_id(user: User, requested: uuid.UUID | None = None) -> uuid.UUID:
    """Crews are pinned to their own team; admins may inspect any team."""
    if user.role == UserRole.ADMIN and requested is not None:
        return requested
    if user.team_id is None:
        raise PermissionDeniedError(
            "Your account is not linked to a repair team. Ask an administrator to assign you to one."
        )
    return user.team_id


def _load_assignment(db: Session, assignment_id: uuid.UUID, user: User) -> RepairAssignment:
    assignment = RepairRepository(db).get_assignment(assignment_id)
    if assignment is None:
        from app.core.errors import NotFoundError

        raise NotFoundError("That job could not be found.")

    if user.role != UserRole.ADMIN and assignment.team_id != user.team_id:
        # Same message as "not found" so ids cannot be probed for existence.
        from app.core.errors import NotFoundError

        raise NotFoundError("That job could not be found.")

    return assignment


@router.get("/dashboard", response_model=TeamDashboardResponse, summary="Crew dashboard")
def team_dashboard(
    team_id: uuid.UUID | None = None,
    user: User = Depends(require_team),
    db: Session = Depends(get_session),
) -> TeamDashboardResponse:
    resolved = _resolve_team_id(user, team_id)
    repository = RepairRepository(db)
    team = repository.get_team(resolved)
    if team is None:
        from app.core.errors import NotFoundError

        raise NotFoundError("That team does not exist.")

    assignments = repository.list_team_assignments(resolved, limit=200)
    now = datetime.now(UTC)
    end_of_day = now.replace(hour=23, minute=59, second=59)

    tasks = [to_task(item) for item in assignments]
    open_tasks = [
        task for task in tasks if task.status in (AssignmentStatus.ASSIGNED, AssignmentStatus.IN_PROGRESS)
    ]

    today = [
        task
        for task in open_tasks
        if task.due_at is None or _utc(task.due_at) <= end_of_day + timedelta(hours=0)
    ]
    critical = [task for task in open_tasks if task.complaint.priority_level == PriorityLevel.CRITICAL]
    in_progress = [task for task in open_tasks if task.status == AssignmentStatus.IN_PROGRESS]
    completed = [
        task
        for task in tasks
        if task.status in (AssignmentStatus.COMPLETED, AssignmentStatus.VERIFIED)
    ][:10]

    return TeamDashboardResponse(
        team=RepairTeamResponse.model_validate(team),
        today=today,
        critical=critical,
        in_progress=in_progress,
        completed_recently=completed,
        stats={
            "open_jobs": len(open_tasks),
            "critical_open": len(critical),
            "in_progress": len(in_progress),
            "overdue": sum(1 for task in open_tasks if task.is_overdue),
            "completed_total": sum(
                1
                for task in tasks
                if task.status in (AssignmentStatus.COMPLETED, AssignmentStatus.VERIFIED)
            ),
            "capacity": team.max_concurrent_jobs,
        },
    )


@router.get("/tasks", response_model=list[TaskResponse], summary="Assigned jobs")
def list_tasks(
    # Without an explicit Query(), FastAPI reads a list-typed parameter from the
    # request body, which a GET does not send - the filter would be silently
    # ignored rather than rejected.
    status_filter: list[AssignmentStatus] | None = Query(None),
    team_id: uuid.UUID | None = Query(None),
    user: User = Depends(require_team),
    db: Session = Depends(get_session),
) -> list[TaskResponse]:
    resolved = _resolve_team_id(user, team_id)
    assignments = RepairRepository(db).list_team_assignments(resolved, statuses=status_filter)
    return [to_task(item) for item in assignments]


@router.get("/tasks/{assignment_id}", response_model=TaskResponse, summary="Job detail")
def get_task(
    assignment_id: uuid.UUID,
    user: User = Depends(require_team),
    db: Session = Depends(get_session),
) -> TaskResponse:
    return to_task(_load_assignment(db, assignment_id, user))


@router.post("/tasks/{assignment_id}/start", response_model=TaskResponse, summary="Start a repair")
def start_task(
    assignment_id: uuid.UUID,
    _: None = Depends(write_rate_limit),
    user: User = Depends(require_team),
    db: Session = Depends(get_session),
) -> TaskResponse:
    assignment = _load_assignment(db, assignment_id, user)
    return to_task(ComplaintService(db).start_repair(assignment, user))


@router.post(
    "/tasks/{assignment_id}/complete",
    response_model=TaskResponse,
    summary="Submit completion evidence",
)
async def complete_task(
    assignment_id: uuid.UUID,
    note: str | None = Form(None),
    photos: list[UploadFile] = File(default_factory=list),
    _: None = Depends(write_rate_limit),
    user: User = Depends(require_team),
    db: Session = Depends(get_session),
) -> TaskResponse:
    assignment = _load_assignment(db, assignment_id, user)

    files: list[tuple[bytes, str, str]] = []
    for upload in photos:
        if not upload.filename:
            continue
        raw = await upload.read()
        if len(raw) > settings.MAX_UPLOAD_BYTES:
            raise ValidationError(
                f"{upload.filename} is too large. Please keep photos under "
                f"{settings.MAX_UPLOAD_BYTES / (1024 * 1024):.0f}MB."
            )
        files.append((raw, upload.content_type or "application/octet-stream", upload.filename))

    updated = ComplaintService(db).complete_repair(
        assignment, user, note=(note or "").strip() or None, evidence_files=files
    )
    return to_task(updated)


def _utc(value: datetime) -> datetime:
    return value if value.tzinfo else value.replace(tzinfo=UTC)

File: supabase/functions/api/routes/team.ts
// Ported from backend/app/api/v1/team.py. A crew member sees only their own
// team's work - the team comes from the authenticated user's profile
// (team_id), never from a request parameter, so one crew cannot read or act
// on another's jobs by guessing an id. Admins may inspect any team.

import { Hono } from "hono";
import { requireTeam, type AuthedProfile } from "../_shared/auth.ts";
import { serviceClient, userClient } from "../_shared/supabase.ts";
import { NotFoundError, PermissionDeniedError, ValidationError } from "../_shared/errors.ts";
import { writeRateLimit } from "../_shared/rate_limit.ts";
import { toTask, toTaskDetail } from "../_shared/serializers.ts";
import { settings } from "../_shared/config.ts";
import { getTeam, listTeamAssignments, getAssignmentFull, type RepairAssignmentRow, type RepairDetails } from "../repositories/repair.ts";
import { completeRepair, escalateAssignment, flagAssignment, startRepair, type EvidenceFile } from "../services/complaints.ts";
import type { AssignmentStatus } from "../_shared/enums.ts";

export const team = new Hono();

function resolveTeamId(user: AuthedProfile, requested: string | null): string {
  if (user.role === "ADMIN" && requested) return requested;
  if (!user.team_id) {
    throw new PermissionDeniedError(
      "Your account is not linked to a repair team. Ask an administrator to assign you to one.",
    );
  }
  return user.team_id;
}

async function loadAssignment(assignmentId: string, user: AuthedProfile): Promise<RepairAssignmentRow> {
  const assignment = await getAssignmentFull(serviceClient(), assignmentId);
  if (!assignment) throw new NotFoundError("That job could not be found.");
  // Same message as "not found" so ids cannot be probed for existence.
  if (user.role !== "ADMIN" && assignment.team_id !== user.team_id) {
    throw new NotFoundError("That job could not be found.");
  }
  return assignment;
}

team.get("/dashboard", async (c) => {
  const req = c.req.raw;
  const user = await requireTeam(req);
  const url = new URL(req.url);
  const resolved = resolveTeamId(user, url.searchParams.get("team_id"));

  const client = serviceClient();
  const repairTeam = await getTeam(client, resolved);
  if (!repairTeam) throw new NotFoundError("That team does not exist.");

  const assignments = await listTeamAssignments(client, resolved, { limit: 200 });
  const now = new Date();
  const endOfDay = new Date(now);
  endOfDay.setUTCHours(23, 59, 59, 999);

  const tasks = assignments.map(toTask);
  const openTasks = tasks.filter((t) => t.status === "ASSIGNED" || t.status === "IN_PROGRESS");
  const today = openTasks.filter((t) => !t.due_at || new Date(t.due_at as string) <= endOfDay);
  const critical = openTasks.filter((t) => (t.complaint as any)?.priority_level === "CRITICAL");
  const inProgress = openTasks.filter((t) => t.status === "IN_PROGRESS");
  const emergencyOpen = openTasks.filter((t) => t.is_emergency);
  const allCompleted = tasks.filter((t) => t.status === "COMPLETED" || t.status === "VERIFIED");

  const durations = allCompleted
    .filter((t) => t.completed_at && t.created_at)
    .map((t) => (new Date(t.completed_at as string).getTime() - new Date(t.created_at as string).getTime()) / 3_600_000);
  const everAssigned = tasks.filter((t) => t.status !== "CANCELLED").length;

  return c.json({
    team: repairTeam,
    today,
    critical,
    in_progress: inProgress,
    completed_recently: allCompleted.slice(0, 10),
    stats: {
      open_jobs: openTasks.length,
      critical_open: critical.length,
      emergency_open: emergencyOpen.length,
      in_progress: inProgress.length,
      overdue: openTasks.filter((t) => t.is_overdue).length,
      completed_total: allCompleted.length,
      capacity: repairTeam.max_concurrent_jobs,
      average_completion_hours: durations.length
        ? Math.round((durations.reduce((a, b) => a + b, 0) / durations.length) * 10) / 10
        : null,
      // Of everything the crew has ever taken on (a flagged/cancelled job was
      // never really "their" work to finish), how much is done - not a
      // point-in-time open/closed split, but the whole track record.
      completion_rate: everAssigned ? Math.round((allCompleted.length / everAssigned) * 1000) / 10 : null,
    },
  });
});

team.get("/tasks", async (c) => {
  const req = c.req.raw;
  const user = await requireTeam(req);
  const url = new URL(req.url);
  const resolved = resolveTeamId(user, url.searchParams.get("team_id"));
  const statuses = url.searchParams.getAll("status_filter") as AssignmentStatus[];

  const assignments = await listTeamAssignments(serviceClient(), resolved, {
    statuses: statuses.length > 0 ? statuses : undefined,
  });
  return c.json(assignments.map(toTask));
});

team.get("/tasks/:id", async (c) => {
  const user = await requireTeam(c.req.raw);
  const assignment = await loadAssignment(c.req.param("id"), user);
  return c.json(toTaskDetail(assignment));
});

team.post("/tasks/:id/start", async (c) => {
  const req = c.req.raw;
  const user = await requireTeam(req);
  await writeRateLimit(userClient(req), req);

  const assignment = await loadAssignment(c.req.param("id"), user);
  const updated = await startRepair(assignment, user);
  return c.json(toTask({ ...updated, complaint: (assignment as any).complaint }));
});

team.post("/tasks/:id/complete", async (c) => {
  const req = c.req.raw;
  const user = await requireTeam(req);
  await writeRateLimit(userClient(req), req);

  const assignment = await loadAssignment(c.req.param("id"), user);

  const form = await req.formData();
  const note = (form.get("note") as string | null)?.trim() || null;
  const photoEntries = form.getAll("photos").filter((f): f is File => f instanceof File && Boolean(f.name));

  const files: EvidenceFile[] = [];
  for (const photo of photoEntries) {
    if (photo.size > settings.MAX_UPLOAD_BYTES) {
      throw new ValidationError(
        `${photo.name} is too large. Please keep photos under ${Math.round(settings.MAX_UPLOAD_BYTES / (1024 * 1024))}MB.`,
      );
    }
    const bytes = new Uint8Array(await photo.arrayBuffer());
    files.push({ bytes, contentType: photo.type || "application/octet-stream" });
  }

  const hadExistingEvidence = ((assignment as any).evidence ?? []).length > 0;
  const updated = await completeRepair(assignment, user, note, files, hadExistingEvidence, repairDetailsFromForm(form));
  return c.json(toTask({ ...updated, complaint: (assignment as any).complaint }));
});

/**
 * The structured "what it took" fields alongside the free-text note - every
 * field optional, so a crew in a hurry can still submit with just a photo
 * and a note, same as before this existed. `null` (not an object of nulls)
 * when nothing was filled in, so `completeRepair` leaves the column alone
 * rather than overwriting a previous submission's details with blanks.
 */
function repairDetailsFromForm(form: FormData): RepairDetails | null {
  const text = (key: string) => (form.get(key) as string | null)?.trim() || null;
  const workersRaw = text("workers_count");
  const workers = workersRaw !== null ? Number(workersRaw) : null;
  const costRaw = text("cost_amount");
  const cost = costRaw !== null ? Number(costRaw) : null;

  const details: RepairDetails = {
    repair_type: text("repair_type"),
    materials: text("materials"),
    quantity: text("quantity"),
    equipment: text("equipment"),
    workers_count: workers !== null && Number.isFinite(workers) && workers >= 0 ? workers : null,
    cost_amount: cost !== null && Number.isFinite(cost) && cost >= 0 ? cost : null,
  };
  return Object.values(details).some((value) => value !== null) ? details : null;
}

team.post("/tasks/:id/flag", async (c) => {
  const req = c.req.raw;
  const user = await requireTeam(req);
  await writeRateLimit(userClient(req), req);

  const assignment = await loadAssignment(c.req.param("id"), user);
  const body = await req.json().catch(() => ({}));
  if (typeof body.reason !== "string" || body.reason.trim().length < 3) {
    throw new ValidationError("Please explain what is wrong with this assignment.");
  }

  const updated = await flagAssignment(assignment, user, body.reason.trim());
  return c.json(toTask({ ...updated, complaint: (assignment as any).complaint }));
});

team.post("/tasks/:id/escalate", async (c) => {
  const req = c.req.raw;
  const user = await requireTeam(req);
  await writeRateLimit(userClient(req), req);

  const assignment = await loadAssignment(c.req.param("id"), user);
  const body = await req.json().catch(() => ({}));
  if (typeof body.reason !== "string" || body.reason.trim().length < 3) {
    throw new ValidationError("Please say what makes this an emergency.");
  }

  const updated = await escalateAssignment(assignment, user, body.reason.trim());
  return c.json(toTask({ ...updated, complaint: (assignment as any).complaint }));
});
